include/aws/s3/private/s3_parallel_read_stream.h

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#ifndef AWS_S3_PARALLEL_READ_STREAM_H
+#define AWS_S3_PARALLEL_READ_STREAM_H
+
+#include <aws/s3/s3.h>
+
+#include <aws/common/ref_count.h>
+
+AWS_PUSH_SANE_WARNING_LEVEL
+
+struct aws_byte_buf;
+struct aws_future_bool;
+struct aws_input_stream;
+
+struct aws_event_loop_group;
+
+struct aws_parallel_input_stream {
+    const struct aws_parallel_input_stream_vtable *vtable;
+    struct aws_allocator *alloc;
+    struct aws_ref_count ref_count;
+
+    void *impl;
+};
+
+struct aws_parallel_input_stream_vtable {
+    /**
+     * Destroy the stream, its refcount has reached 0.
+     */
+    void (*destroy)(struct aws_parallel_input_stream *stream);
+
+    /**
+     * Read into the buffer in parallel.
+     * The implementation needs to support this to be invoked concurrently from multiple threads
+     *
+     * TODO: error handling:
+     * - The stream reach end before all parts
+     * - Failed to create a elg task to read
+     * - Asyc related stuff as the original async stream (If we cannot fill the dest, we wait?)
+     * - etc.
+     */
+    struct aws_future_bool *(*read)(
```

Comment:
why is this async? it does force use to do an extra allocation
why not just synchronous like aws_input_stream?

Reply:
async is more generic. I would assume people want to do parallel read from all sorts of things, and some may not be a synchronous call. 

Although, yeah, if we do open the file on needed, we don't need it to be async, but I think make it async is more future prove. 

Also, we can make it async if we need to, but I don't think this will add complexity now, as we already have code path to handle async stream, and well, an extra allocation I guess not really matter